### ISSUE TITLE
Proposal: Tekton Workflows

### ISSUE THREAD

**Issue by dibyom** [tektoncd/community#464]
This is a proposal for a higher level resource that would group the lower level Tekton primitives into a single entity. Conceptually, this would be similar to workflow in [GitHub Actions](https://docs.github.com/en/actions/quickstart#creating-your-first-workflow) and [Argo Workflows](https://argoproj.github.io/projects/argo/), and Pipeline in [Jenkins](https://www.jenkins.io/doc/book/pipeline/syntax/#triggers), [CircleCI](https://circleci.com/docs/2.0/pipelines/#overview), and [Gitlab](https://docs.gitlab.com/ee/ci/pipelines/).

### The problem the project will solve

The main goal is to simplify the CI/CD configuration for end users. Tekton today has different un-opinionated primitives. While this is very flexible, it adds additional burden on end users looking for simpler CI/CD use cases. Platform developers tend to abstract away this complexity by adding abstraction layers that only expose basic inputs to the app developers.

Within Tekton itself, there have been a few experiments in this area. The [Generators](https://github.com/tektoncd/experimental/tree/main/generators) project combines trigger and pipeline definitions into a higher level opinionated syntax that is aware of external systems such as GitHub. The [Pipelines as Code](https://github.com/openshift-pipelines/pipelines-as-code#event-matching-to-a-pipeline) project uses PipelineRuns with annotations to describe Triggers. Similarly, [Tekton Integrations](https://github.com/tektoncd/community/pull/147) proposed adding annotations for notifications and status updates to external systems. There was also a proposal for a concept of "[Project](https://github.com/tektoncd/dashboard/pull/1612)" that will combine Triggers and Pipelines into a single view. 

Similarly, many other CI/CD systems have an abstraction where primitives such as Triggers, and Pipelines are combined into one resource e.g. workflow in [GitHub Actions](https://docs.github.com/en/actions/quickstart#creating-your-first-workflow) and [Argo Workflows](https://argoproj.github.io/projects/argo/), and Pipeline in [Jenkins](https://www.jenkins.io/doc/book/pipeline/syntax/#triggers), [CircleCI](https://circleci.com/docs/2.0/pipelines/#overview), and [Gitlab](https://docs.gitlab.com/ee/ci/pipelines/).

## Who Will Own It

@dibyom
@bobcatfish

**Comment by vdemeester** [comment]
@dibyom @bobcatfish few questions 
- Would this be an *opiniated* setup/use of the different TektonCD components ?
- Should we aim to be the solution we use for dogfooding or not ? (cc @afrittoli)
- Can this be started as an experimental project ? If not (or even if), should we have a TEP for it ? If we go experimental, I wouldn't require it, if we want to make it a project on its own, I would require it I think.

**Comment by nikhil-thomas** [comment]
@nikhil-thomas

**Comment by dibyom** [comment]
> Would this be an opiniated setup/use of the different TektonCD components ?

Yes, some similarities with https://github.com/openshift-pipelines/pipelines-as-code

> Should we aim to be the solution we use for dogfooding or not ? (cc @afrittoli)

Absolutely, dogfooding would be a great use case!

> Can this be started as an experimental project ? If not (or even if), should we have a TEP for it ? If we go experimental, I wouldn't require it, if we want to make it a project on its own, I would require it I think.

Yes! this is for an experimental project but I do plan on opening up a TEP so that we can discuss the actual API in more detail

> How does it relates to "Pipelines as Code", or others experimentation in that sense ? Also, would it be a full "distribution" (aka you install it, it installs everything) or a component that sits on top of the others and let the "installation/configuration" part up to the user (or something else, like the operator 🙃 )


Good question -- pipelines as code also groups triggers + other options together  (using annotations). So in that sense it is similar. The concept of repo CRD is also very interesting -- would be interesting to explore a generic eventSource type thing perhaps for Workflows. I think this would be less of a distribution and more of a component (something like the operator could do the install/configuration).

I think the pipelines as code WG that @chmouel proposed would be a good forum to explore these questions.

**Comment by dibyom** [comment]
According to the [Proposing Projects process](https://github.com/tektoncd/community/blob/main/process.md#proposing-projects), we need two governing board members to approve before we can create an experimental project. 

/cc @vdemeester @afrittoli @abayer @bobcatfish

**Comment by bobcatfish** [comment]
👍  from me!

**Comment by vdemeester** [comment]
👍🏼

**Comment by dibyom** [comment]
Awesome! I'll go ahead with setting up the experimental folder!

**Comment by dibyom** [comment]
Status Update - we started a [Working Group](https://github.com/tektoncd/community/blob/main/working-groups.md#workflows) leading to a [proposed design](https://docs.google.com/document/d/1CaEti30nnq95jd-bD1Iep4mEnZ5qMEBHax0ixq-9xP0/edit?resourcekey=0-rbRAlTIqe9jnavyIQK_5lQ#) (join [mailing list](https://github.com/tektoncd/community/blob/main/contact.md#mailing-list) for access) and an ongoing experimental implementation in https://github.com/tektoncd/experimental/tree/main/workflows

**Comment by vdemeester** [comment]
Closing — the experimental repository is frozen and this TEP was never implemented. With Pipelines-as-Code moving into the `tektoncd` org, the "pipeline definitions along the code" use case is addressed differently. Removing from the [Pipeline CI Dogfooding project](https://github.com/orgs/tektoncd/projects/29).
